Custodia-related changes

mbarnes opened the pull request:
Just miscellaneous tweaks to support the Custodia feature in `commissaire-service`.

~~**[WIP]** until I'm sure I have everything needed for `commissaire` in one PR.~~

mbarnes committed:
Vagrantfile: Use the custodia pip package instead of rpm.

The rpm is out-of-date and custodia is now a hard dependency of
commissaire-service.

However the pip package doesn't include any systemd files, so supply our
own custodia.service that runs immediately (bypassing socket activation).

Also threw in a hack that will hopefully make systemd services come up
more reliably on startup.

mbarnes committed:
storage: Remove unused StoreHandlerBase._get_connection().

Vestige of an earlier design.

mbarnes committed:
models: Add SecretModel._key_container attribute.

Each subclass chooses a unique Custodia key container name.

e.g. HostCreds stores secrets in a key container named "hosts".

mbarnes committed:
Test model class attribute invariants.

Make sure required class attributes are populated.

mbarnes commented:
@ashcrow  So just an update on this...

`commissaire-http` requires the `py-bcrypt` package (via `pip`), but apparently one of Custodia's requirements pulls in `bcrypt` (still trying to hunt that down).  Both packages install a module called `bcrypt`, and `commissaire-http` is now importing the `bcrypt` package's module.

There are slight incompatibilities between the two which are breaking E2E tests.  See the latest [travis-ci failures](https://travis-ci.org/projectatomic/commissaire/builds/242880089?utm_source=github_status&utm_medium=notification).

We switched to `py-bcrypt` because, at the time, it was packaged for Fedora and `bcrypt` wasn't.  But that doesn't seem as relevant anymore.  So I'm wondering if, instead of fighting it, we should reverse https://github.com/projectatomic/commissaire-mvp/pull/40.

ashcrow commented:
@mbarnes yeah, it's worth reverting rather than porting custodia over to ``py-bcrypt``. It should be pretty simple to change. TBH I like ``bcrypt`` better anyway 😄

mbarnes committed:
Vagrantfile: Simplify custodia authentication.

In a production environment, the Custodia instance would be running
inside the storage-service container and be inaccessible from outside
the container.  So we can relax authentication and just have Custodia
look for a REMOTE_USER header.

mbarnes committed:
Deal with custodia directory in etcd.

Created when needed, destroy when needed.

mbarnes committed:
e2e: Update for Custodia integration.

Yet another service to start...

As with the other services, a private Custodia instance can be started
for the duration of the test suite with:

   behave -D start-custodia

or

   behave -D start-all-servers

Also, the step "a host already exists at {host}" requires talking
directly to Custodia's REST API instead of through Commissaire.  This in
turn requires a special requests.Session with UNIX Domain Socket support
as in CustodiaStoreHandler.

mbarnes committed:
tox.ini: Add "-D start-custodia" to behave command.

mbarnes commented:
@ashcrow  I've been fighting with Travis CI for a couple days now trying to get the test environment set up properly.  The same `behave` command Travis is using works fine for me locally, so I'm removing the **WIP** tag now so you can start reviewing this.  Could you also verify E2E tests work on your machine?

I want to keep this PR open until Travis CI works, but I thought in having you review it now you might spot some test environment issue I overlooked.

ashcrow commented:
Looks like significant CI breakage around authentication (403's). Expected?

mbarnes commented:
Not expected.  I can reproduce it locally now with the `tox` command, but running `behave` directly works fine for some reason.  Probably something stupid in `environment.py`.

ashcrow commented:
OK. I'm running ``vagrant`` to verify that first.

ashcrow commented:
Using this PR up to 13cd5430557638d34fa7e14c0927a6e996094de9 locally with vagrant I get:

```
      AttributeError: 'Context' object has no attribute 'CUSTODIA_SOCKET_URL'
```

Starting at

```
  Scenario: Recreating a compatible host with authentication  # features/hosts/create.feature:35
```

I see the code sets it but it seems to not be actually doing it in time?

ashcrow commented:
The latest patch (e64ab091d420b32f13e48c336a07c55d9bfcb895)is better. Now I see this:

```
        File."../commissaire/src/commissaire/util/unixadapter.py", line 48, in connect
          sock.connect(socket_path)
      requests.packages.urllib3.exceptions.ProtocolError: ('Connection aborted.', FileNotFoundError(2, 'No such file or directory'))
```

ashcrow commented:
I'm guessing it's a race condition with the ``socket_path``
```
def start_custodia(context, args):
    """Starts a Custodia instance."""
    context.CUSTODIA_DATA_DIR = tempfile.TemporaryDirectory()
    tmpdir = context.CUSTODIA_DATA_DIR.name

    socket_path = os.path.join(tmpdir, 'socket', 'custodia.sock')
    socket_url = 'http+unix://' + quote(socket_path, safe='')
```

In before all

```
    # Will be overwritten if using 'start-custodia'.
    socket_path = '/var/run/custodia/custodia.sock'
    socket_url = 'http+unix://' + quote(socket_path, safe='')
    context.CUSTODIA_SOCKET_PATH = socket_path
    context.CUSTODIA_SOCKET_URL = socket_url
```

ashcrow commented:
It might be work making the testing sock location in temp always and using the custodia configuration template to provide the temp dir.

mbarnes commented:
I realized this afternoon running these tests with Vagrant is a problem, because [in one of the steps to silently set up a host as a precondition](https://github.com/projectatomic/commissaire/pull/141/commits/4df1b03df6b4da97c0066b71b0832000dfe6122f#diff-82673601740f602008f7bfde7b3312d4), it writes directly to `custodia.sock`.  Problem is, that socket only exists on the Vagrant VM, not on the host machine where the tests are running.

Going to have to either put the `custodia.sock` in a shared directory, or see if Custodia can listen to both a socket file and an HTTP port at the same time (so tests have a back channel to Custodia).

In any case, `tools/behave` without Vagrant is working fine for me.

ashcrow commented:
I'll give it a go locally today and report back.

mbarnes commented:
Boo.  😞  Guess it'll have to use a shared directory.
```
        if url and sock:
            raise ValueError(
                "'server_url' and 'server_socket' are mutually exclusive.")
```
https://github.com/latchset/custodia/blob/master/src/custodia/server/config.py#L126

ashcrow commented:
I get the following error running ``tools/behave`` with this PR applied:

```
FileNotFoundError: [Errno 2] No such file or directory: '/tmp/tmprvey9kfl/lib/master.key'
```

ashcrow commented:
The error is because the directories don't exist in the temp dir ...something like this in ``start_custodia`` made it get further:
```
    # Make the internal custodia directories
    for d in ('lib', 'log', 'run', 'socket'):
        os.mkdir(os.path.sep.join([tmpdir, d]))

    context.PROCESSES['custodia'] = subprocess.Popen(['custodia', conffile])
    time.sleep(1)
    context.PROCESSES['custodia'].poll()
```

ashcrow commented:
``WebOb`` needs to be added to requirements as well as it's needed by our use of custodia.

https://pypi.python.org/pypi/WebOb/

WebOb #MIT

ashcrow commented:
With the above changes I get:
```
12 features passed, 0 failed, 0 skipped
51 scenarios passed, 0 failed, 10 skipped
294 steps passed, 0 failed, 75 skipped, 0 undefined
```

😄

mbarnes commented:
> The error is because the directories don't exist in the temp dir ...

Hmm, that's weird.  The inline Custodia config in `environment.py` should have `makedirs = True`.  ~~Although, Custodia's implementation of that option is poor because it seems to error out if the directories already exist.  Should be idempotent.~~

**Ed**: Nevermind, I'm wrong about that part.  `makedirs` *is* idempotent.

mbarnes commented:
Thanks @ashcrow, I'll add your recommendations as well as my own realizations and try to get this all sorted out today.

mbarnes commented:
> WebOb needs to be added to requirements as well as it's needed by our use of custodia.

@ashcrow, where did you read that?  I can't find any reference to WebOb in the Custodia sources.

ashcrow commented:
@mbarnes that was the last error I got when running the tests, that webob's middleware was not installed (though custodia was). Once I added webob things worked.

mbarnes commented:
Okay, verified that after uninstalling WebOb.

mbarnes commented:
> Boo. 😞 Guess it'll have to use a shared directory.

Turns out `sshfs` doesn't support UNIX domain sockets, so putting the socket in a shared directory won't work either.

Plan C: run two `custodia` instances -- one on the Vagrant VM (for Commissaire) and one on the host machine (for test steps) -- configured to use the same `master.key` and speak to the same `etcd` server.

Oye vey!

ashcrow commented:
@mbarnes what about running custodia and and storage-service on the local host with all other items on the vagrant boxes?

mbarnes commented:
> @mbarnes what about running custodia and and storage-service on the local host with all other items on the vagrant boxes?

Well, the Vagrant environment is also useful for manual testing, so I'd prefer it to remain complete and not have missing pieces.  There's value in having a fully functioning test environment with simply `vagrant up`.

mbarnes commented:
Latest `fixup!` gets `tools/behave -D use-vagrant` working.

Still getting constant 403 errors through `tox` or Travis CI.

mbarnes committed:
Don't require commctl in test environments.

commctl still depends on py-bcrypt, but commissaire-http recently
switched back to bcrypt.  The problem is, both packages install a
module named "bcrypt" and when py-bcrypt is installed it tends to
win in "import bcrypt" calls.  The two APIs are not compatible.

The test environments don't do anything with commctl, so leave it
out.

mbarnes committed:
bdd-requirements.txt: Remove disabled lines.

Clearer without the commented out URLs.

mbarnes committed:
tox.ini: Fix command for "bdd" test environment.

mbarnes commented:
@ashcrow  All set (*finally!*).  The key to the test failures was https://github.com/projectatomic/commissaire/pull/141/commits/29abc84d2084522d0fbe2c2c21889a203c4bb4bb.

ashcrow commented:
![yes](https://i.imgflip.com/n0gny.jpg)

ashcrow commented on features/environment.py:
```diff
@@ -345,7 +421,23 @@ def before_all(context):
+        config_custodia(context)
+        libdir = os.path.join(context.CUSTODIA_TMPDIR.name, 'lib')
+        os.mkdir(libdir, mode=0o700)
+        os.symlink(
```
👍 good idea

ashcrow commented on vagrant/custodia/custodia.conf:
```diff
@@ -1,13 +1,8 @@
-libdir = /var/lib/custodia
-logdir = /var/log/custodia
-rundir = /var/run/custodia
+shareddir = /vagrant/commissaire/vagrant/custodia
```
👍 good idea

ashcrow reviewed (commented):
LGTM
